File: agent/query_engine/EngineEvent.py
from __future__ import annotations

import json
import time
from dataclasses import dataclass
from typing import Any


@dataclass(frozen=True)
class EngineEvent:
    event: str
    source: str
    payload: dict[str, Any]
    trace_id: str | None = None
    event_version: str = "1.0"

    @classmethod
    def llm_delta(cls, text: str, *, trace_id: str | None = None) -> "EngineEvent":
        return cls(event="llm_data", source="llm", payload={"text": text}, trace_id=trace_id)

    @classmethod
    def tool_use(
        cls,
        *,
        tool_name: str,
        tool_call_id: str,
        input_payload: dict[str, Any],
        trace_id: str | None = None,
    ) -> "EngineEvent":
        return cls(
            event="tool_use",
            source="tool",
            payload={"tool_name": tool_name, "tool_call_id": tool_call_id, "input": input_payload},
            trace_id=trace_id,
        )

    @classmethod
    def tool_result(cls, payload: dict[str, Any], *, trace_id: str | None = None) -> "EngineEvent":
        return cls(event="tool_result", source="tool", payload=payload, trace_id=trace_id)

    @classmethod
    def tool_error(cls, payload: dict[str, Any], *, trace_id: str | None = None) -> "EngineEvent":
        return cls(event="tool_error", source="tool", payload=payload, trace_id=trace_id)

    @classmethod
    def sys_done(cls, finish_reason: str, *, trace_id: str | None = None) -> "EngineEvent":
        return cls(event="sys_done", source="system", payload={"finish_reason": finish_reason}, trace_id=trace_id)

    @classmethod
    def sys_error(
        cls,
        *,
        code: str,
        message: str,
        retryable: bool,
        trace_id: str | None = None,
    ) -> "EngineEvent":
        return cls(
            event="sys_error",
            source="system",
            payload={"code": code, "message": message, "retryable": retryable},
            trace_id=trace_id,
        )

    def to_sse(self) -> str:
        data = {
            "event_version": self.event_version,
            "trace_id": self.trace_id or "",
            "timestamp": int(time.time() * 1000),
            "source": self.source,
            "payload": self.payload,
        }
        return f"event: {self.event}\ndata: {json.dumps(data, ensure_ascii=False)}\n\n"

File: agent-ts/src/query_engine/EngineEvent.ts
import type { JsonObject } from "../common/json/types/JsonTypes.js";

export interface EngineEventProps {
  readonly event: string;
  readonly source: string;
  readonly payload: JsonObject;
  readonly trace_id?: string | null;
  readonly event_version?: string;
}

export class EngineEvent {
  constructor(
    public readonly event: string,
    public readonly source: string,
    public readonly payload: JsonObject,
    public readonly trace_id: string | null = null,
    public readonly event_version: string = "1.0"
  ) {}

  get traceId(): string | null {
    return this.trace_id;
  }

  get eventVersion(): string {
    return this.event_version;
  }

  static llmDelta(text: string, traceId?: string | null): EngineEvent {
    return new EngineEvent("llm_data", "llm", { text }, traceId ?? null);
  }

  static toolUse(
    toolName: string,
    toolCallId: string,
    inputPayload: JsonObject,
    traceId?: string | null
  ): EngineEvent {
    return new EngineEvent(
      "tool_use",
      "tool",
      { tool_name: toolName, tool_call_id: toolCallId, input: inputPayload },
      traceId ?? null
    );
  }

  static toolResult(payload: JsonObject, traceId?: string | null): EngineEvent {
    return new EngineEvent("tool_result", "tool", payload, traceId ?? null);
  }

  static toolError(payload: JsonObject, traceId?: string | null): EngineEvent {
    return new EngineEvent("tool_error", "tool", payload, traceId ?? null);
  }

  static sysDone(finishReason: string, traceId?: string | null): EngineEvent {
    return new EngineEvent("sys_done", "system", { finish_reason: finishReason }, traceId ?? null);
  }

  static sysError(code: string, message: string, retryable: boolean, traceId?: string | null): EngineEvent {
    return new EngineEvent("sys_error", "system", { code, message, retryable }, traceId ?? null);
  }

  toSse(): string {
    const body = {
      event_version: this.event_version,
      trace_id: this.trace_id ?? "",
      timestamp: Date.now(),
      source: this.source,
      payload: this.payload
    };
    return `event: ${this.event}\ndata: ${JSON.stringify(body)}\n\n`;
  }
}
